### ISSUE TITLE
SSBC-EM2: Customers can use alpha version of SSBC

### ISSUE THREAD

**Issue by mrnugget** [sourcegraph/sourcegraph#24421]
Items:

- [x] `executor-queue` either needs to be (a) merged into `frontend` or (b) part of our standard deployments so its available for customers
  - [x] https://github.com/sourcegraph/sourcegraph/pull/22873
- [ ] One executor job per workspace (repository+path+commit)
  - [ ] src-cli able to execute a batch spec in a given workspace (not doing resolving of workspaces at all) @mrnugget
    - [ ] Last log message is the rendered `changeset_spec` (it doesn't create it)
  - [ ] Add `createPendingBatchSpec` mutation @mrnugget
    - [ ] Merge `batch_spec` & `batch_spec_executions` tables
      - Extend `batch_spec` table first, and only 
    - [ ] Create `batch_spec_workspace_execution_jobs` table
      - has workspace details: repo, rev, path
      - executor-job details: state, started_at, execution_logs, ...
      - batch_spec_id, changeset_spec_id
    - [ ] In mutation: create `batch_spec`, which is a dbworker-job, that kicks off workspace-resolution
    - [ ] Each workspace yielded by the execution of this job creates a "pending" `batch_spec_workspace_execution_job`
    - [ ] User can now preview which workspaces would be executed
  - [ ] Execute `batch_spec_workspace_execution_job`, one per executor @mrnugget
    - [ ] `batch_spec_workspace_execution_job` are executor-jobs etc.
    - [ ] Each workspace-execution-job, if successful, creates a `changeset_spec` (from the JSON log line of `src-cli`) with correct `batch_spec_id` and stores `changeset_spec.id` on `batch_spec_workspace_execution_job.changeset_spec_id`
  - [ ] Build and switch frontend UI to use new API @eseliger
    - [ ] Add `batchSpec.workspaces` to GraphQL API
      - Has repository/workspace information
      - It's an executor job
      - Has log output, like executor_job now
    - [ ] GraphQL API allows starting of all jobs belonging to batch spec
    - [ ] Move parsing of log lines to backend and expose proper status API
    - [ ] Build workflow to create, preview and then run a batch spec
  - [ ] Clean up non-executed batch specs
      - [ ] Lazy cleanup in backend (every x days) @mrnugget
      - [ ] Cleanup previous `batch_specs` from frontend if user creates new one and didn't execute previous one @eseliger
- [ ] Workflow improvements: 
  - [ ] Retry with slightly adjusted spec input after error
  - [ ] Good “create batch spec” workflow, make it look good when no batch spec exists yet.
- [ ] Deployment of `executor` @eseliger
  - [ ] Improve deployment of executor
    - [x] Build AMI in CI for executor https://github.com/sourcegraph/sourcegraph/pull/23364
    - [ ] Deploy new executor from CI
    - [x] Generalize code-intel terraform stuff in infra repo https://github.com/sourcegraph/infrastructure/pull/2704
  - [ ] Documentation for customers on how to deploy `executor` @eseliger
    - [ ] Bare metal
    - [ ] Google Cloud
    - [ ] AWS
  - [ ] Publish terraform files for `executor` on Google Cloud
  - [ ] Publish terraform files for `executor` on AWS
- [ ] Get rid of technical debt in authentication tokens before shipping
  - [ ] https://github.com/sourcegraph/sourcegraph/issues/22763
- [ ] Testing
  - [ ] Long-running batch specs (1000 repositories) work reliably
  - [ ] Storybook entries for execution page components
  - [ ] Check if any newly introduced resolvers are untested

**Comment by malomarrec** [comment]
Our likely first alpha tester is running Sourcegraph in AWS.

**Comment by eseliger** [comment]
Good news, I am almost done reaching feature parity of AWS and GCP.

**Comment by malomarrec** [comment]
@courier-new @eseliger @mrnugget it sounds like we added "Build UI components that don't rely on data model" in this iteration. I suggest moving it to a separate issue, to make it clear that it is out of scope for EM2, and non-blocking (but very important to set us up for success in the _next_ iteration).

**Why?**

At this stage, it's more important that we are able to ship something to the customer fast than improve the look and feel.
What we want to accomplish here is 
> As a customer, I want to run a large batch changes, that I can parallelize on several executors so that they run faster. I'm fine with debugging with the CLI first. All I want to do is paste a spec, run it across executors, be able to see the logs per repo/step, then go to preview/apply. I also want to be able to edit the spec and re-run it to correct mistakes.

The current look and feel is actually fine for shipping this, because the key thing we want to show is "you compute one batch change over several executors, and it's faster".

**WDYT?**

**Comment by mrnugget** [comment]
Sounds good to me. We can move it into a separate ticket and list it on the #22764 after EM2 and then, once we know what we'll do after EM2, we can re-sort it.

**Comment by malomarrec** [comment]
There is a risk that EM2 is blocked by executor deployment work that Code Intel is doing. If that's the case:
- we can help / support Code Intel if helpful to them
- the outcome of EM2 becomes: "As a Sourcegraph team member I can run a large batch change, that I can parallelize on several executors so that they run faster."

**Comment by mrnugget** [comment]
@courier-new just FYI after @malomarrec @eseliger and I discussed the above comments in a call, we moved the two blocks of tasks assigned to you out of this EM to #22764. Doesn't change the work, nor its priority, it's just bookkeeping, a change of representation to make sure that reaching EM2 is not blocked or dependent on the work you're doing, but that the next milestones will be.
